hydroflow_lang/src/graph/ops/source_json.rs

```diff
@@ -0,0 +1,76 @@
+use quote::quote_spanned;
+
+use crate::graph::{OpInstGenerics, OperatorInstance};
+
+use super::{
+    FlowProperties, FlowPropertyVal, OperatorConstraints, OperatorWriteOutput, WriteContextArgs,
+    RANGE_0, RANGE_1,
+};
+
+/// > 0 input streams, 1 output stream
+///
+/// > Arguments: An [`AsRef`](https://doc.rust-lang.org/std/convert/trait.AsRef.html)`<`[`Path`](https://doc.rust-lang.org/nightly/std/path/struct.Path.html)`>`
+/// for a file to read as json. This will emit the parsed value one time.
+///
+/// `source_json` may take one generic type argument, the type of the value to be parsed, which must implement [`Deserialize`](https://docs.rs/serde/latest/serde/de/trait.Deserialize.html).
+///
+/// ```hydroflow
+/// source_json("example.json") -> for_each(|json: hydroflow::serde_json::Value| println!("{:#?}", json));
+/// ```
+#[hydroflow_internalmacro::operator_docgen]
+pub const SOURCE_JSON: OperatorConstraints = OperatorConstraints {
+    name: "source_json",
+    hard_range_inn: RANGE_0,
+    soft_range_inn: RANGE_0,
+    hard_range_out: RANGE_1,
+    soft_range_out: RANGE_1,
+    num_args: 1,
+    persistence_args: RANGE_0,
+    type_args: &(0..=1),
+    is_external_input: true,
+    ports_inn: None,
+    ports_out: None,
+    properties: FlowProperties {
+        deterministic: FlowPropertyVal::Yes,
+        monotonic: FlowPropertyVal::Yes,
+        inconsistency_tainted: false,
+    },
+    input_delaytype_fn: |_| None,
+    write_fn: |wc @ &WriteContextArgs {
+                   root,
+                   op_span,
+                   ident,
+                   op_inst:
+                       OperatorInstance {
+                           generics: OpInstGenerics { type_args, .. },
+                           arguments,
+                           ..
+                       },
+                   ..
+               },
+               _| {
+        let generic_type = type_args
+            .get(0)
+            .map(quote::ToTokens::to_token_stream)
+            .unwrap_or(quote_spanned!(op_span=> _));
+
+        let ident_jsonread = wc.make_ident("jsonread");
+        let write_prologue = quote_spanned! {op_span=>
+            // Note that reading the entire file to memory is (probably still) faster than using a
+            // reader: https://github.com/serde-rs/json/issues/160
+            let mut #ident_jsonread = {
+                let string = ::std::fs::read_to_string(#arguments).unwrap();
+                let value: #generic_type = #root::serde_json::from_str(&*string).unwrap();
```

Comment:
maybe tokio::fs here so it doesn't block the thread?

Reply:
Yeahh the bad news is that `serde/serde_json` doesn't support async at all ... partially because json is a terrible format for streaming..

Reply:
json lines `JSONL` would work though ..

Reply:
Oh my comment was on the wrong line. I meant the std::fs::read part, we could make that async. It might not really matter though since this only happens once on startup and it would be a bunch of effort to add async to that. So probably no reason to block this PR for now, but 

Streaming is another concern and I agree there's not really anything you can do there because json is not a good format for streaming.